src/number_theory/arithmetic_function.lean

```diff
@@ -0,0 +1,256 @@
+by cases f; cases g; cases h; refl
+
+@[ext] theorem ext ⦃f g : arithmetic_function α⦄ (h : ∀ x, f x = g x) : f = g :=
+coe_inj (funext h)
+
+theorem ext_iff {f g : arithmetic_function α} : f = g ↔ ∀ x, f x = g x :=
+⟨λ h x, h ▸ rfl, λ h, ext h⟩
+
+@[simp]
+lemma map_zero {f : arithmetic_function α} : f 0 = 0 := f.map_zero'
+
+@[simp]
+lemma zero_apply {x : ℕ} : (0 : arithmetic_function α) x = 0 := rfl
+
+section has_one
+variable [has_one α]
+
+instance : has_one (arithmetic_function α) := ⟨⟨λ x, ite (x = 1) 1 0, rfl⟩⟩
+
+@[simp]
+lemma one_one : (1 : arithmetic_function α) 1 = 1 := rfl
+
+@[simp]
+lemma one_apply_of_ne_one {x : ℕ} (h : x ≠ 1) : (1 : arithmetic_function α) x = 0 := if_neg h
+
+end has_one
+end has_zero
+
+instance nat_coe [semiring α] : has_coe (arithmetic_function ℕ) (arithmetic_function α) :=
+⟨λ f, ⟨↑(f : ℕ → ℕ), by { transitivity ↑(f 0), refl, simp }⟩⟩
+
+@[simp]
+lemma nat_coe_apply [semiring α] {f : arithmetic_function ℕ} {x : ℕ} :
+  (f : arithmetic_function α) x = f x := rfl
+
+instance int_coe [ring α] : has_coe (arithmetic_function ℤ) (arithmetic_function α) :=
+⟨λ f, ⟨↑(f : ℕ → ℤ), by { transitivity ↑(f 0), refl, simp }⟩⟩
+
+@[simp]
+lemma int_coe_apply [ring α] {f : arithmetic_function ℤ} {x : ℕ} :
+  (f : arithmetic_function α) x = f x := rfl
+
+@[simp]
+lemma coe_coe [ring α] {f : arithmetic_function ℕ} :
+  ((f : arithmetic_function ℤ) : arithmetic_function α) = f :=
+by { ext, simp, }
+
+section add_monoid
+
+variable [add_monoid α]
+
+instance : has_add (arithmetic_function α) := ⟨λ x y, ⟨λ n, x n + y n, by simp⟩⟩
+
+@[simp]
+lemma add_apply {f g : arithmetic_function α} {n : ℕ} : (f + g) n = f n + g n := rfl
+
+instance : add_monoid (arithmetic_function α) :=
+{ add_assoc := λ _ _ _, ext (λ _, add_assoc _ _ _),
+  zero_add := λ _, ext (λ _, zero_add _),
+  add_zero := λ _, ext (λ _, add_zero _),
+  .. (infer_instance : has_zero (arithmetic_function α)),
+  .. (infer_instance : has_add (arithmetic_function α)) }
+
+end add_monoid
+
+instance [add_comm_monoid α] : add_comm_monoid (arithmetic_function α) :=
+{ add_comm := λ _ _, ext (λ _, add_comm _ _),
+  .. (infer_instance : add_monoid (arithmetic_function α)) }
+
+instance [add_group α] : add_group (arithmetic_function α) :=
+{ neg := λ f, ⟨λ n, - f n, by simp⟩,
+  add_left_neg := λ _, ext (λ _, add_left_neg _),
+  .. (infer_instance : add_monoid (arithmetic_function α)) }
+
+instance [add_comm_group α] : add_comm_group (arithmetic_function α) :=
+{ .. (infer_instance : add_comm_monoid (arithmetic_function α)),
+  .. (infer_instance : add_group (arithmetic_function α)) }
+
+section dirichlet_ring
+variable [semiring α]
+
+/-- The Dirichlet convolution of two arithmetic functions `f` and `g` is another arithmetic function
+  such that `(f * g) n` is the sum of `f x * g y` over all `(x,y)` such that `x * y = n`. -/
+def convolve (f g : arithmetic_function α) : arithmetic_function α :=
+⟨λ n, ∑ x in divisors_antidiagonal n, f x.fst * g x.snd, by simp⟩
+
```

Comment:
Do we want `convolve` or `(*)` to be the standard notation? I'm inclined to support the latter, since it will unlock the `semiring` simp set.

In that case, I would define the `has_mul` instance here:
```suggestion
/-- The Dirichlet convolution of two arithmetic functions `f` and `g` is another arithmetic function
  such that `(f * g) n` is the sum of `f x * g y` over all `(x,y)` such that `x * y = n`. -/
instance : has_mul (arithmetic_function α) :=
⟨λ f g, ⟨λ n, ∑ x in divisors_antidiagonal n, f x.fst * g x.snd, by simp⟩⟩
```

(and update the remainder of the file accordingly).